src/main/scala/chisel3/util/SparseVec.scala

```diff
@@ -0,0 +1,383 @@
+          ) ++ elements.values).toSeq
+        )
+
+        // Build up a sequence of BitPats that map the addresses to the encoding
+        // in the denseVec.  If out-of-bounds values are possible and the user
+        // instructed us to return the first-element on out-of-bounds acesses,
+        // then fill out the array with BitPats to do this.
+        val bitPats = indices.zipWithIndex.map {
+          case (index, i) =>
+            BitPat(index.U(addrWidth.W)) -> BitPat(d.encoding(i + baseEncodedAddress, encodedSize))
+        } ++ {
+          (outOfBoundsValue, zeroValue) match {
+            case (OutOfBoundsBehavior.Indeterminate, _) | (_, None) => Seq.empty
+            case (OutOfBoundsBehavior.First, Some(_)) =>
+              (size until BigInt(addrWidth).pow(2).toInt).map {
+                case i =>
+                  BitPat(i.U(addrWidth.W)) -> BitPat(d.encoding(baseEncodedAddress, encodedSize))
+              }
+          }
+        }
+
+        // Generate the truth table.  If this SparseVec does not have a default
+        // value or if the DefaultValue is indeterminate, then treat the default
+        // value as don't care---any value indexed into the dense vector is
+        // fine.  Otherwise, the default value exists and is always encoded as
+        // address zero.
+        val ttable = TruthTable(
+          bitPats,
+          (hasDefaultValue, defaultValue) match {
+            case (false, _) | (_, DefaultValueBehavior.Indeterminate) => BitPat.dontCare(encodedSize)
+            case _                                                    => BitPat(d.encoding(0, encodedSize))
+          }
+        )
+
+        // Use a decoder to generate a lookup into the denseVec using the
+        // provided Lookup.Decoder strategy.  Return the value.
+        d.lookup(
+          decoder(addr, ttable),
+          denseVec
+        ).asInstanceOf[A]
+
+      // Generate a lookup into the elements of the SparseVec using when
+      // statements.
+      case Lookup.IfElse =>
+        // The result of the lookup.
+        val result = Wire(gen)
+
+        // The default value _must_ be specified to bypass initialization checking.
+        result := defaultValue.getValue(lookupType)
+
+        // Generate one when statement for each value.
+        indices.zip(elements.values).foreach {
+          case (index, data) =>
+            when(addr === index.U) {
+              result := data
+            }
+        }
+
+        // If the elements have a value at index zero and if the user indicated
+        // they want to return this value on out-of-bounds behavior, then return
+        // it.  Otherwise, add no logic for this case.
+        (outOfBoundsValue, zeroValue) match {
+          case (SparseVec.OutOfBoundsBehavior.Indeterminate, _) | (_, None) =>
+          case (SparseVec.OutOfBoundsBehavior.First, Some(data)) =>
+            when(addr >= size.U) {
+              result := data
+            }
+        }
```

Comment:
I don't think this logic is correct when you have `SparseVec.OutOfBoundsBehavior.First` combined with no element at index `0`. In order to be equivalent to a dense Vec, we need to give a "fake" zero-th value which effectively is the default `value`. Basically, I think this can be fixed as follows:

```suggestion
        outOfBoundsValue match {
          case SparseVec.OutOfBoundsBehavior.Indeterminate =>
          case SparseVec.OutOfBoundsBehavior.First, Some(data)) =>
            val oob = zeroValue.getOrElse(defaultValue.getValue(lookupType)) // assign result of this expression to a val up above
            when(addr >= size.U) {
              result := oob
            }
        }
```

This could use a test. This case I'm describing here came up in the real design when I deployed this same basic idea (but in a much less reusable way, this PR is awesome).

Reply:
I added a test case. However, this seems to pass. I think this logic is correct for the reason that the `result` is already set, outside of any when, to `defaultValue.getValue(lookupType)`. Hence, the PR should only need to set the value to the `zeroValue` and just let it fallback to the default.

Does that make sense?